In [2]:
import pandas as pd
import numpy as np
from pathlib import Path

In [3]:
# Path to the raw data folder
DATA_PATH = Path("../data/raw")

# Get all CSV files
csv_files = sorted(DATA_PATH.glob("*.csv"))

# Check the number of CSV files
print(f"Number of CSV files: {len(csv_files)}")

# Load all CSV files into a dictionary
data = {}

for file in csv_files:
    df_name = file.stem
    data[df_name] = pd.read_csv(file)

    print(f"Loaded: {file.name}")

Number of CSV files: 7


C:\Users\Sarvesh\AppData\Local\Temp\ipykernel_8216\1111105226.py:15: DtypeWarning: Columns (6) have mixed types. Specify dtype option on import or set low_memory=False.
  data[df_name] = pd.read_csv(file)


Loaded: inventory_movements.csv
Loaded: inventory_opening_balances.csv
Loaded: products.csv
Loaded: purchase_order_lines.csv
Loaded: purchase_orders.csv
Loaded: suppliers.csv
Loaded: warehouses.csv


In [4]:
dataset_summary = pd.DataFrame([
    {
        "Dataset": name,
        "Rows": df.shape[0],
        "Columns": df.shape[1]
    }
    for name, df in data.items()
])

dataset_summary

,Dataset,Rows,Columns
0,inventory_movements,138436,8
1,inventory_opening_balances,30000,5
2,products,2500,6
3,purchase_order_lines,53967,6
4,purchase_orders,18000,7
5,suppliers,250,5
6,warehouses,12,4


In [5]:
for name, df in data.items():
    print(f"\n{'=' * 60}")
    print(f"Dataset: {name}")
    print(f"{'=' * 60}")

    display(df.head())


Dataset: inventory_movements


,movement_id,product_id,warehouse_id,movement_at,movement_type,quantity_change,purchase_order_line_id,transfer_id
0,M00000001,IP00117,W010,2023-10-26,receipt,51,POL0000001,NaN
1,M00000002,IP01414,W012,2025-04-25,receipt,484,POL0000002,NaN
2,M00000003,IP02164,W012,2025-04-25,receipt,326,POL0000003,NaN
3,M00000004,IP00664,W012,2025-04-25,receipt,85,POL0000004,NaN
4,M00000005,IP02414,W012,2025-04-25,receipt,339,POL0000005,NaN



Dataset: inventory_opening_balances


,opening_balance_id,product_id,warehouse_id,balance_date,opening_units
0,IP00001-W001,IP00001,W001,2023-01-01,446
1,IP00001-W002,IP00001,W002,2023-01-01,171
2,IP00001-W003,IP00001,W003,2023-01-01,377
3,IP00001-W004,IP00001,W004,2023-01-01,338
4,IP00001-W005,IP00001,W005,2023-01-01,149



Dataset: products


,product_id,supplier_id,sku,category,unit_cost,reorder_point
0,IP00001,SUP0001,SC-000001,components,539.77,350
1,IP00002,SUP0002,SC-000002,packaging,430.74,397
2,IP00003,SUP0003,SC-000003,finished_goods,772.73,65
3,IP00004,SUP0004,SC-000004,consumables,562.48,491
4,IP00005,SUP0005,SC-000005,spares,372.97,246



Dataset: purchase_order_lines


,purchase_order_line_id,purchase_order_id,product_id,quantity_ordered,quantity_received,unit_cost
0,POL0000001,PO000001,IP00117,130,51,766.23
1,POL0000002,PO000002,IP01414,484,484,184.48
2,POL0000003,PO000002,IP02164,326,326,254.58
3,POL0000004,PO000002,IP00664,85,85,127.93
4,POL0000005,PO000002,IP02414,339,339,284.95



Dataset: purchase_orders


,purchase_order_id,supplier_id,warehouse_id,ordered_at,expected_at,received_at,status
0,PO000001,SUP0117,W010,2023-10-05,2023-10-25,2023-10-26,partial
1,PO000002,SUP0164,W012,2025-03-01,2025-04-26,2025-04-25,received
2,PO000003,SUP0135,W002,2024-12-17,2025-02-11,2025-02-13,received
3,PO000004,SUP0186,W006,2023-07-06,2023-08-25,2023-09-01,received
4,PO000005,SUP0131,W007,2025-12-13,2025-12-31,NaN,open



Dataset: suppliers


,supplier_id,supplier_name,country_code,lead_time_days,reliability_score
0,SUP0001,Supplier 001,CN,38,0.925
1,SUP0002,Supplier 002,CA,50,0.940
2,SUP0003,Supplier 003,US,34,0.842
3,SUP0004,Supplier 004,CA,11,0.973
4,SUP0005,Supplier 005,DE,23,0.903



Dataset: warehouses


,warehouse_id,warehouse_name,region,capacity_units
0,W001,West DC 01,West,175970
1,W002,Southwest DC 02,Southwest,238544
2,W003,Midwest DC 03,Midwest,221995
3,W004,Northeast DC 04,Northeast,259097
4,W005,Southeast DC 05,Southeast,83700


In [6]:
for name, df in data.items():
    print(f"\n{'=' * 60}")
    print(f"Dataset: {name}")
    print(f"{'=' * 60}")

    df.info()


Dataset: inventory_movements
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 138436 entries, 0 to 138435
Data columns (total 8 columns):
 #   Column                  Non-Null Count   Dtype 
---  ------                  --------------   ----- 
 0   movement_id             138436 non-null  object
 1   product_id              138436 non-null  object
 2   warehouse_id            138436 non-null  object
 3   movement_at             138436 non-null  object
 4   movement_type           138436 non-null  object
 5   quantity_change         138436 non-null  int64 
 6   purchase_order_line_id  51369 non-null   object
 7   transfer_id             24134 non-null   object
dtypes: int64(1), object(7)
memory usage: 8.4+ MB

Dataset: inventory_opening_balances
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   opening_balance_id  30000 non-null

In [7]:
for name, df in data.items():
    missing = pd.DataFrame({
        "Missing Values": df.isnull().sum(),
        "Missing Percentage": (
            df.isnull().sum() / len(df) * 100
        ).round(2)
    })

    missing = missing[missing["Missing Values"] > 0]

    print(f"\n{'=' * 60}")
    print(f"Missing Values: {name}")
    print(f"{'=' * 60}")

    if missing.empty:
        print("No missing values found.")
    else:
        display(missing)


Missing Values: inventory_movements


,Missing Values,Missing Percentage
purchase_order_line_id,87067,62.89
transfer_id,114302,82.57



Missing Values: inventory_opening_balances
No missing values found.

Missing Values: products
No missing values found.

Missing Values: purchase_order_lines
No missing values found.

Missing Values: purchase_orders


,Missing Values,Missing Percentage
received_at,873,4.85



Missing Values: suppliers
No missing values found.

Missing Values: warehouses
No missing values found.


In [8]:
for name, df in data.items():
    duplicate_count = df.duplicated().sum()

    print(f"{name}: {duplicate_count} duplicate rows")

inventory_movements: 0 duplicate rows
inventory_opening_balances: 0 duplicate rows
products: 0 duplicate rows
purchase_order_lines: 0 duplicate rows
purchase_orders: 0 duplicate rows
suppliers: 0 duplicate rows
warehouses: 0 duplicate rows


In [9]:
for name, df in data.items():
    print(f"\n{name}")
    print(list(df.columns))


inventory_movements
['movement_id', 'product_id', 'warehouse_id', 'movement_at', 'movement_type', 'quantity_change', 'purchase_order_line_id', 'transfer_id']

inventory_opening_balances
['opening_balance_id', 'product_id', 'warehouse_id', 'balance_date', 'opening_units']

products
['product_id', 'supplier_id', 'sku', 'category', 'unit_cost', 'reorder_point']

purchase_order_lines
['purchase_order_line_id', 'purchase_order_id', 'product_id', 'quantity_ordered', 'quantity_received', 'unit_cost']

purchase_orders
['purchase_order_id', 'supplier_id', 'warehouse_id', 'ordered_at', 'expected_at', 'received_at', 'status']

suppliers
['supplier_id', 'supplier_name', 'country_code', 'lead_time_days', 'reliability_score']

warehouses
['warehouse_id', 'warehouse_name', 'region', 'capacity_units']


In [10]:
# Check movement types and their frequency
data["inventory_movements"]["movement_type"].value_counts(
    dropna=False
)

# Investigate missing IDs by movement type
movement_missing_analysis = (
    data["inventory_movements"]
    .groupby("movement_type", dropna=False)
    .agg(
        Total_Records=("movement_id", "count"),
        Missing_PO_Line_ID=("purchase_order_line_id",
                            lambda x: x.isna().sum()),
        Missing_Transfer_ID=("transfer_id",
                             lambda x: x.isna().sum())
    )
    .reset_index()
)

movement_missing_analysis


,movement_type,Total_Records,Missing_PO_Line_ID,Missing_Transfer_ID
0,adjustment,7492,7492,7492
1,receipt,51369,0,51369
2,sale,55441,55441,55441
3,transfer,24134,24134,0


In [11]:
# Check purchase order statuses
data["purchase_orders"]["status"].value_counts(
    dropna=False
)

# Compare received_at values with order status
received_date_analysis = (
    data["purchase_orders"]
    .groupby("status", dropna=False)
    .agg(
        Total_Orders=("purchase_order_id", "count"),
        Missing_Received_At=("received_at",
                             lambda x: x.isna().sum())
    )
    .reset_index()
)

received_date_analysis["Missing_Percentage"] = (
    received_date_analysis["Missing_Received_At"]
    / received_date_analysis["Total_Orders"]
    * 100
).round(2)

received_date_analysis

,status,Total_Orders,Missing_Received_At,Missing_Percentage
0,cancelled,406,406,100.0
1,open,467,467,100.0
2,partial,1326,0,0.0
3,received,15801,0,0.0


In [12]:
# Compare total rows, non-null IDs, and movement types
movements = data["inventory_movements"]

print("Total rows:", len(movements))
print("Non-null movement IDs:", movements["movement_id"].count())
print("Missing movement IDs:", movements["movement_id"].isna().sum())
print("Missing movement types:", movements["movement_type"].isna().sum())

# Count rows per movement type, including rows with missing values
movements.groupby(
    "movement_type",
    dropna=False
).size().reset_index(name="Row_Count")

Total rows: 138436
Non-null movement IDs: 138436
Missing movement IDs: 0
Missing movement types: 0


,movement_type,Row_Count
0,adjustment,7492
1,receipt,51369
2,sale,55441
3,transfer,24134


## Key Findings & Initial Insights

### 1. Dataset Overview
- The project contains seven CSV datasets with a combined total of 243,165 rows and 41 columns.
- The datasets cover inventory movements, opening inventory balances, products, purchase order lines, purchase orders, suppliers, and warehouses.
- `inventory_movements` is the largest dataset, containing 138,436 records.

### 2. Missing Values
- Five datasets contain no missing values: `inventory_opening_balances`, `products`, `purchase_order_lines`, `suppliers`, and `warehouses`.
- In `inventory_movements`, `purchase_order_line_id` is missing for adjustment, sale, and transfer movements. `transfer_id` is missing for adjustment, receipt, and sale movements.
- The missing inventory movement IDs appear to follow the movement type. These values should be retained unless further validation identifies a data-quality issue.
- In `purchase_orders`, 873 records (4.85%) have a missing `received_at` value. All 406 cancelled orders and 467 open orders have no receipt date, while partial and received orders have a receipt date.
- The missing receipt dates appear consistent with order status and should not be filled arbitrarily.

### 3. Initial Relationship Observations
- Products can be linked to suppliers using `supplier_id`.
- Purchase order lines can be linked to purchase orders and products using `purchase_order_id` and `product_id`.
- Purchase orders contain supplier and warehouse identifiers.
- Inventory movements and opening balances contain product and warehouse identifiers, making it possible to analyse inventory by product and warehouse.
- Inventory movements may also link to purchase order lines or transfers, depending on the movement type.
- These relationships require validation before analysis.

### 4. Potential Areas for Analysis
- Inventory movement patterns by movement type, product, and warehouse.
- Purchase order fulfilment and outstanding orders.
- Supplier lead times and reliability.
- Inventory levels compared with product reorder points.
- Warehouse-level inventory distribution and capacity.

### 5. Next Steps
- Validate data types and convert date columns where necessary.
- Check duplicate records and uniqueness of identifier columns.
- Validate foreign-key relationships between datasets.
- Investigate date ranges, quantity values, and other potential data-quality issues.
- Document the cleaning decisions before proceeding to exploratory data analysis.
